not calculate new priority when user update other spec of a pod

CaoShuFeng opened the pull request:
motivation of this change:
If we update the priority, pod validation mechanism will prevent this update
request, this is not expected.

**Release note**:
```
Priority admission controller: not calculate new priority when user update a pod
```

CaoShuFeng commented:
/assign @bsalamat

bsalamat commented on plugin/pkg/admission/priority/admission_test.go:
```diff
@@ -414,3 +414,175 @@ func TestPodAdmission(t *testing.T) {
 	}
 }
+
+func TestUpdatePodAdmission(t *testing.T) {
```
I think this test is overly complex for testing update and does not test "Update". All that update does is to leave integer value of priority intact if PodSpec.PriorityClassName has not changed.
So, the test scenario should be like this:
1. Create a PriorityClass.
2. Create a pod with PriorityClassName set to the name of the priority class you created in step 1.
3. Delete the PriorityClass and add another one with the same name, but a different value.
4. Update the pod without changing its PriorityClassNaem.
5. Make sure that the integer value of Priority in PodSpec has not changed.

CaoShuFeng replied:
Done.

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws

bsalamat commented on plugin/pkg/admission/priority/admission_test.go:
```diff
@@ -32,7 +32,7 @@ import (
 )
 
-func addPriorityClasses(ctrl *PriorityPlugin, priorityClasses []*scheduling.PriorityClass) {
+func setPriorityClasses(ctrl *PriorityPlugin, priorityClasses []*scheduling.PriorityClass) {
```
Any particular reason for renaming this function? I think "add" was a better name.

CaoShuFeng replied:
No strong reasons for this renaming.
For me:
`add` means keep both the original classes and the new classes.
`keep` means, remove the original classes, then add the new classes.

`add` is also OK for me.

CaoShuFeng replied:
OK, I have taken the name back.

bsalamat reviewed (commented):
Thanks, @CaoShuFeng! Looks good. Just a minor comment regarding renaming a function.

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws

bsalamat commented:
/lgtm
/approve no-issue

bsalamat commented:
/retest

deads2k commented on plugin/pkg/admission/priority/admission.go:
```diff
@@ -164,6 +164,16 @@ func (p *PriorityPlugin) admitPod(a admission.Attributes) error {
+			}
+			if oldPod.Spec.PriorityClassName == pod.Spec.PriorityClassName {
+				// if user didn't modify priority in purpose, we don't caculate the new one.
+				return nil
```
This doesn't look right to me.  What is stopping me from leaving priorityclassname alone and changing priority directly?

deads2k replied:
> This doesn't look right to me. What is stopping me from leaving priorityclassname alone and changing priority directly?

Oh, and if that is a hole, add a test that checks for it and ensures you can't bypass it like that.

bsalamat replied:
Our pod update validation logic allows only several fields in PodSpec to be updated. Priority and PriorityClassName cannot be updated:
https://github.com/bsalamat/kubernetes/blob/master/pkg/api/validation/validation.go#L2910

deads2k replied:
> Our pod update validation logic allows only several fields in PodSpec to be updated. Priority and PriorityClassName cannot be updated:
> https://github.com/bsalamat/kubernetes/blob/master/pkg/api/validation/validation.go#L2910

Wait.  If you can't update the value here, why are you checking for any updates at all?

CaoShuFeng replied:
> Wait. If you can't update the value here, why are you checking for any updates at all?

What about:
When user tries to update PriorityClassName or Priority, we return admission.NewForbidden directly?

bsalamat replied:
@deads2k As @CaoShuFeng mentioned, the problem here is that if someone updated a field of a Pod, this admission controller would try to resolve the Priority of the Pod again. Now if the PriorityClass was changed and had a different value, the admission controller would update the pod spec with the new value causing an error being thrown by the validation logic. This PR fixes this scenario by not changing the integer value of the pod during updates.

deads2k replied:
> @deads2k As @CaoShuFeng mentioned, the problem here is that if someone updated a field of a Pod, this admission controller would try to resolve the Priority of the Pod again. Now if the PriorityClass was changed and had a different value, the admission controller would update the pod spec with the new value causing an error being thrown by the validation logic. This PR fixes this scenario by not changing the integer value of the pod during updates.

But I think that's all this plugin does.  Why not just stop running any logic here for updates at all?  Around line 117 or so.

CaoShuFeng replied:
> But I think that's all this part of the plugin does. Why not just stop running any logic here for updates at all? Around line 117 or so.

I am OK with this. the validation logic will return status code 422 anyway.
What do you think? @bsalamat

bsalamat replied:
> But I think that's all this part of the plugin does. Why not just stop running any logic here for updates at all? Around line 117 or so.

I agree. The historical reason for having update monitoring was that we first thought we would allow updating "PriorityClassName" of pods, but we decided not to support it. We can remove the update monitoring.

CaoShuFeng replied:
Done.

CaoShuFeng committed:
not calculate new priority when user update pods

motivation of this change:
If we update the priority, pod validation mechanism will prevent this update
request, this is not expected.

bsalamat commented on plugin/pkg/admission/priority/admission_test.go:
```diff
@@ -414,3 +414,76 @@ func TestPodAdmission(t *testing.T) {
 	}
 }
+
+func TestUpdatePodAdmission(t *testing.T) {
```
With the new changes and the fact that we no longer use this admission controller when a pod is updated, I am not sure if this test is relevant any more. I think we can drop the test.

CaoShuFeng replied:
Done.

CaoShuFeng commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

deads2k commented:
lgtm - will leave tagging for @bsalamat 

/approve no-issue

bsalamat commented:
Thanks a lot, @CaoShuFeng and @deads2k!

/lgtm
